lib/language_pack/helpers/outdated_ruby_version.rb

```diff
@@ -0,0 +1,150 @@
+# Queries S3 in the background to determine
+# what versions are supported so they can be recommended
+# to the user
+#
+# Example:
+#
+#   ruby_version = LanguagePack::RubyVersion.new("ruby-2.2.5")
+#   outdated = LanguagePack::Helpers::OutdatedRubyVersion.new(
+#     current_ruby_version: ruby_version,
+#     fetcher: LanguagePack::Fetcher.new(LanguagePack::Base::VENDOR_URL, "heroku-16")
+#   )
+#
+#   outdated.call
+#   puts outdated.suggested_ruby_minor_version
+#   #=> "ruby-2.2.10"
+class LanguagePack::Helpers::OutdatedRubyVersion
+  DEFAULT_RANGE = 1..5
+  attr_reader :current_ruby_version
+
+  def initialize(current_ruby_version: , fetcher:)
+    @current_ruby_version = current_ruby_version
+    @fetcher      = fetcher
+
+    @minor_versions = [current_ruby_version]
+    @eol_versions = []
+    @threads = []
+  end
+
+  # Enqueues checks in the background
+  def call
+    return unless can_check?
+
+    check_minor_versions
+    check_eol_versions_major
+    check_eol_versions_minor
+    self
+  end
+
+  def join
+    return false if current_ruby_version.patchlevel_is_significant?
+    return false if current_ruby_version.rbx?
+    return false if current_ruby_version.jruby?
+
+    @threads.each(&:join)
```

Comment:
how much does doing this in parallel actually save? It creates a lot to think about (and a lot that could go wrong)

Reply:
It's a non-trivial amount of time if we're doing it in serial. Here's a benchmark script you can run on heroku:

```
$ heroku run bash -a heroku-buildpack-ruby
~$ irb -I ./lib

require 'language_pack'
ENV["STACK"] = "heroku-18"
pack = LanguagePack::Ruby.new(".", "/dev/null")
fetcher = pack.instance_variable_get("@fetchers")[:mri]

exists_time = Benchmark.measure do
  fetcher.exists?("ruby-2.4.0.tgz")
end.real

does_not_exist_time = Benchmark.measure do
  fetcher.exists?("ruby-2.4.0asdfasdfasdf.tgz")
end.real

raw_time  = Benchmark.measure do
`bash -c set -o pipefail; curl -L --fail --retry 5 --retry-delay 1 --connect-timeout 3 --max-time 30 -I https://s3-external-1.amazonaws.com/heroku-buildpack-ruby/heroku-18/ruby-2.4.0.tgz`
end.real

puts "Time to check binary that exists:         #{exists_time}"
puts "Time to check binary that DOES NOT exist: #{does_not_exist_time}"
```


```
Time to check binary that exists:         0.1962961310055107
Time to check binary that DOES NOT exist: 0.2097533739870414
```

So if we're checking 5+ urls then we're adding a extra second to build time. These results are way worse locally since S3 may be far away, the worst I saw was 1.5 seconds for the case where the binary does not exist. Technically that shouldn't be a concern, but fast tests locally is pretty nice.

Reply:
I would personally probably eat the seconds, and do it serial. But I also don't see anything wrong here. It's read-only right?

Reply:
> It's read-only right? 

These threads do write to data structures, but they're doing so in threadsafe ways with MRI and i'm purposefully not firing this logic on jruby. My last class at georgia tech involved doing a lot of parallel computation in C, i'm reasonably confident in the threadsafety here and we have tests that exercise all of this logic.